Add reference counter for platform context

RonEld opened the pull request:
### Description
1. Move the `mbedtls_platform_context` to be platform code, in `features/mbedtls/platfrom/`.
2. Add static refernce counter, to setup and teardown the platform code only once.
3. Adjust Cryptocell porting accordingly.

Should work with system modules using Mbed TLS, such as BLE, that wouldn't terminate the cryptographic driver for the application, in the middle of operation.
Adresses first part of #7069 

CC @sbutcher-arm @Patater @pan- 

Question: Should we protect the reference counter with a mutex?

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [x] Fix
    [ ] Refactor
    [ ] New target
    [ ] Feature
    [ ] Breaking change

Commit:
Add reference counter for platform context

1. Move the `mbedtls_platform_context` to be platform code, in `features/mbedtls/platfrom/`.
2. Add static refernce counter, to setup and teardown the platform code only once.
3. Adjust Cryptocell porting accordingly.

simonbutcher commented on features/mbedtls/platform/src/platform_alt.c:
```diff
@@ -0,0 +1,59 @@
+{
+    int ret = 0;
+    if( ctx == NULL )
+        return ( -1 );
```
I don't think we should be returning -1, however, we have failed to define in Mbed TLS any error codes for function.

Maybe we should define error codes now locally in Mbed OS, and upstream them to Mbed TLS in time for the following release of Mbed OS? Possibly `INVALID_INPUT` for the case above, and `HW_FAILURE` in case of failure from the accelerator. What do you think?

RonEld replied:
I agree, but as you said, there is currently no error codes available

simonbutcher replied:
You could define some locally in Mbed OS, and also submit a PR to upstream the same codes. 

As and when the version of Mbed TLS with the new codes becomes available, we can remove the local definitions. 

We control both sides (upstream and downstream) so there's no risk of reusing the same values.

RonEld replied:
I created https://github.com/ARMmbed/mbedtls/pull/1993 to add the platform error code

pan- reviewed (changes requested):
Hi Ron, 

Thanks for this submission that intend to solves #7069. Unfortunately _reference 
counting_ is not the right tool to solve the issue; not at the mbed tls level. 

Reference counting implies that a single instance/context is shared across multiple 
parts of the program. In our case, it is expected that application code using mbed 
tls submit a different/specific context every time `mbedtls_platform_setup` is 
called. 

With the code as it is proposed, a context will never be initialized if their is at least 
another context initialized. This can have dramatic consequences if `crypto_platform_ctx` 
actually contains valuable informations like CRYS_RND_State_t: The code may setup one 
instance and teardown one that hasn't been initialized. 

I strongly believe that keeping track of initialization is an implementation detail
and this as no place in the platform code at mbed tls level.

simonbutcher commented on features/mbedtls/platform/src/platform_alt.c:
```diff
@@ -0,0 +1,59 @@
+    if( reference_count == 1 )
+    {
+        /* call platform specific code to setup crypto driver*/
+        ret = crypto_platform_setup( &ctx->platform_impl_ctx );
```
This makes `crypto_platform_setup()` part of the API for crypto drivers. That will need documenting in the Mbed OS Porting Guide, and also as an interface change will need wider review.

Is there anyway of avoiding this interface change? I'm not sure there is, but we should consider alternatives if we can.

RonEld replied:
I agree it needs documenting in the handbook.

As for interface change, it's basically only a change in the existing name of the function ( `cc_platform_setup` renamed to `crypto+platform_setup`, and a bit of who is responsible of implementing CC initialization has changed). Although it is an API change, I believe it is a minor one
Unfortunately, I don't see any other way to have initialization of the platform,  in an opaque way, and make it portable to other APIs. ( we can have the whole `mbedtls_platform_setup()` functino implemented in driver code, but then we guideusers to implement the reference count , which is very common to all the drivers )

simonbutcher commented on features/mbedtls/platform/src/platform_alt.c:
```diff
@@ -0,0 +1,59 @@
+    if( ctx == NULL )
+        return;
+
+    if( reference_count == 0 )
```
I don't think you need two `if` statements, and it looks a bit odd.

Have you considered using a less than operator? eg.
```
    reference_count--;
    if( reference_count <= 0 )
    {
        crypto_platform_teardown();
    }
```

RonEld replied:
I agree it looks a bit odd, but what you are suggesting is prone to errors:
1. `crypto_platform_teardown()` *might* be called twice, if `reference_count` is 0 or -1.
2. corner cases where we get the reference counter to -2, and then when some chooses to setup, it won't initialize the platform, because the counter will be increased to -1.

Perhaps assigning `reference_count` to 0 after calling `crypto_platform_teardown()`?

simonbutcher replied:
I agree, it would be safer to reassign `reference_count`.

So...
```
    reference_count--;
    if( reference_count <= 0 )
    {
        crypto_platform_teardown();
        reference_count = 0;
    }
```

simonbutcher reviewed (commented):
I think we need to consider the interface changes carefully before we make this change.

cc: @Patater

RonEld commented:
@pan- The idea in the mbedtls layer, is the *platform* itslef is shared across all the modules.
It's basically telling: " start cryptography operations" and "stop cryptography operations".

Specifically for NRF53840, there shouldn't be a problem, as the driver is initialized once, and terminated once. The `CRYS_RND_State_t` holds the trng data for seeding the DRBG. It is used in the CC initialization because the API requires it, but in actual usage, this struct is used in the trng function, for gathering the entropy.

I agree there might be a need for additional tweaking, such as protecting the `CRYS_RND_State_t` in `trng_api.c` with a mutex, and protecting the reference coutner with a mutex, or maybe additional changes, but I think it could be postponed,

pan- commented:
> The idea in the mbedtls layer, is the platform itslef is shared across all the modules.

Thanks for the clarification. 

In that case wouldn't it make sense to remove `crypto_platform_ctx` from `mbedtls_platform_context` and declare a single instance of `crypto_platform_ctx` alongside `reference_count` ? That instance would be the only one passed to `crypto_platform_setup` and `crypto_platform_teardown`. It would prevent calling teardown on an uninitialized `crypto_platform_ctx`.

simonbutcher commented:
Hi @pan- 

That's not really how we intended the platform context to be used. It's not well documented, mainly because no one has provided an implementation until now, but the context is supposed to be a one-per-platform singleton.

The intention was you should initialise once.

The implication of what you're suggesting is that the API we're providing isn't adequate, and you actually need a `mbedtls_platform_setup()`, that takes no context as parameter (as you can't track them), and possibly returns one instead?

Commit:
Add Mbed TLS Platform module errors

1. Add error codes for platform setup \ teardown.
2. Reassign `reference_count` to 0 after terminating platform,
and remove condition for 0

RonEld commented:
@sbutcher-arm I addressed your comments

pan- commented:
Hi @sbutcher-arm 

I agree that if the context passed into `mbedtls_platform_setup` and `mbedtls_platform_teardown` is a singleton then _reference counting_ is an  adequate solution. Thanks for the clarification.

Given that in the current design the platform context singleton lives outside mbed tls platform layer; library and system code API using mbed tls must be updated to accept the platform context singleton instantiated by the application. In some cases - like BLE - that would force us to expose  implementation details that are target specific in our abstract interfaces. 

I wonder if it wouldn't be easier to encapsulate the platform context singleton into mbed tls platform layer and completely hide its existence and management from the API exposed to applications. In other words remove the context parameter in setup and teardown.

What do you think ? 

PS: If the context is a singleton it may be better to add a reference counter field in the platform context structure rather than having another singleton in the global scope.

0xc0170 commented:
@RonEld What is the status of this PR? Are there any outstanding issues to be resolved (the changes requested in the review for instance) ?

RonEld commented:
@0xc0170 I have addressed all the issues. The current status is that there is an architectural decision on how the API should actually look. 
I can't really do much until the decision will be made.

cmonr commented:
@pan- @sbutcher-arm @RonEld Are there conversations happening offline somewhere, or has discussion on this PR simply become stale?

cmonr commented:
Since no progress has been made on this PR for over two weeks, and there's no sign of external conversations, I'm going to close this PR.

Once updates are available, it can be opened.

Commit:
Make the platform context a global variable

Make the platform context a global variable,
adding the refernce counter to it.

RonEld commented:
@cmonr I have available update for this PR. Do you want me to create a new PR or should this one be opened?

cmonr commented:
@RonEld You can commit it to this PR and it can be reopened.

RonEld commented:
@pan- I have updated the context to be a global one, holding the reference counter within. You can now call the `mbedtls_platform_setup()` and `mbedtls_platform_teardown()` with a `NULL` argument, as it's being ignored for Mbed OS now.
Note I haven't made it exactly a singleton ( not static ), as it needs to be accessed by other modules. For example, by the trng module, in cryptocell.
Does this fit your needs?

RonEld commented:
cc @sbutcher-arm @Patater

simonbutcher commented on features/mbedtls/platform/src/platform_alt.c:
```diff
@@ -20,38 +20,33 @@
-
-
-int mbedtls_platform_setup( mbedtls_platform_context *ctx )
+int mbedtls_platform_setup( mbedtls_platform_context *obsolete_ctx )
```
Why is the context called obsolete? I think that's misleading. It's simply unused.

RonEld replied:
It's obsolete, because it's not used and ignored anymore. I can rename

simonbutcher replied:
If it were obsolete it would no longer be used in any context - but that's not true. We're just choosing not to use it in Mbed OS. It *could* be used in other OS's and platforms.

I really think it's misleading still. Thanks for changing it.

simonbutcher commented on features/mbedtls/platform/src/platform_alt.c:
```diff
@@ -20,38 +20,33 @@
-
-
-int mbedtls_platform_setup( mbedtls_platform_context *ctx )
+int mbedtls_platform_setup( mbedtls_platform_context *obsolete_ctx )
```
Should we check `obsolete_ctx` is `NULL`?

RonEld replied:
I don't think so, it is just being ignored. This way, it will be backwards compatible to applications \ modules using this function, that are still sending a content as parameter. Note the example applications weren't updated yet with `NULL` as parameter

simonbutcher commented on features/cryptocell/FEATURE_CRYPTOCELL310/trng.c:
```diff
@@ -22,9 +22,10 @@
 
-extern CRYS_RND_State_t     rndState;
-extern CRYS_RND_WorkBuff_t  rndWorkBuff;
+extern mbedtls_platform_context  ctx;
```
Is this still needed?

RonEld replied:
yes. Since there isn't any `getter` function for the context, we still need access to members of it. The RNG functions need the `rndState`, which is now member of the context

Patater commented on features/cryptocell/FEATURE_CRYPTOCELL310/TARGET_MCU_NRF52840/crypto_platform.c:
```diff
@@ -20,14 +20,25 @@
+#if defined(MBEDTLS_PLATFORM_SETUP_TEARDOWN_ALT)
 
-int cc_platform_setup( cc_platform_ctx *ctx )
+static CRYS_RND_WorkBuff_t  rndWorkBuff = { { 0 } } ;
```
Should this go in the `crypto_platform_ctx`?

RonEld replied:
I don't think so, it's only a sandbox for the initialization operation, and not used later

Patater replied:
OK

Commit:
Rename error codes

1. Rename error codes to fit Mbed TLS error code names.
2. Remove the Invalid input error code, as it's not used anymore.

Patater commented on features/mbedtls/platform/src/platform_alt.c:
```diff
@@ -20,38 +20,33 @@
+    if( ctx.reference_count <= 0 )
+    {
+        /* call platform specific code to terminate crypto driver*/
+        crypto_platform_terminate( &ctx.platform_impl_ctx );
```
This can call terminate multiple times if a client calls teardown too often. Do we care? Maybe this should only call terminate when reference_count goes from 1 to 0, and does nothing otherwise.

We might also want to use atomic decrement if we care about overlapping calls to `mbedtls_platform_teardown()`.

RonEld replied:
This is what I did in the begining. See https://github.com/ARMmbed/mbed-os/pull/7099#discussion_r192667729

>Do we care?

I don't think so, because basically it's an on\off button, us pressing the off button more than once. The likelihood for that is minor, as we do set it to 0 after termination

Patater commented on features/mbedtls/platform/src/platform_alt.c:
```diff
@@ -20,38 +20,33 @@
-    /* call platform specific code to setup CC driver*/
-    if( ( ret = cc_platform_setup( &ctx->platform_impl_ctx ) ) != 0 )
-        return ( ret );
+    ctx.reference_count++;
```
Does this function need to work in multi-threaded environments (multiple callers from different threads)?

If so, consider using a mutex or atomic increment (`core_util_atomic_incr_u32()`). Otherwise, we risk calling setup multiple times if two threads overlap calls to `mbedtls_platform_setup()`. Both threads can read reference count of 0 into a register, increment the register copy, and write back the register to memory at the same time. Then both will proceed to call `crypto_platform_setup()`.

RonEld replied:
I did leave an open question in the Pr description, whether we should consider concurrency issues;
>Question: Should we protect the reference counter with a mutex?

I'll change to atomic increase and decrease

Patater commented on features/mbedtls/platform/src/platform_alt.c:
```diff
@@ -20,38 +20,33 @@
-    return ( 0 );
+    if( ctx.reference_count == 1 )
+    {
+        /* call platform specific code to setup crypto driver*/
```
Add a space after `driver`.

Patater commented on features/mbedtls/platform/src/platform_alt.c:
```diff
@@ -20,38 +20,33 @@
+
+    if( ctx.reference_count <= 0 )
+    {
+        /* call platform specific code to terminate crypto driver*/
```
Add a space after `driver`.

Patater commented on features/cryptocell/FEATURE_CRYPTOCELL310/trng.c:
```diff
@@ -22,9 +22,10 @@
-extern CRYS_RND_State_t     rndState;
-extern CRYS_RND_WorkBuff_t  rndWorkBuff;
+extern mbedtls_platform_context  ctx;
+static CRYS_RND_WorkBuff_t  rndWorkBuff;
```
Should `rndWorkBuff` be part of mbedtls_platform_context?

RonEld replied:
As mentioned in https://github.com/ARMmbed/mbed-os/pull/7099#discussion_r213934349, this is only a sandbox for the initialization operation, and not used later, so I don't think it should be part of the context

Patater commented on features/mbedtls/platform/inc/platform_alt.h:
```diff
@@ -20,50 +20,48 @@
 mbedtls_platform_context;
 
 
+void mbedtls_platform_init( mbedtls_platform_context* ctx);
```
Where did this function declaration come from? Do we need it here? Where are the Doxygen comments for it? Why isn't `mbedtls_platform_teardown()` in this header as well?

In the header file that documents `mbedtls_platform_setup()` and `mbedtls_platform_teardown()`, we should comment as to their thread-safety. (As in, if it is safe to call the functions concurrently from multiple threads.)

RonEld replied:
`mbedtls_platform_init()` is a residue from previous work, it should be removed

RonEld replied:
>In the header file that documents mbedtls_platform_setup() and mbedtls_platform_teardown(), we should comment as to their thread-safety. (As in, if it is safe to call the functions concurrently from multiple threads.)

Probably, but this should be done as a PR in the Mbed TLS repository

RonEld replied:
`mbedtls_platform_setup()` and `mbedtls_platform_teardown()` are part of the Mbed TLS platform.h API, and described there

Commit:
Address concurrency and style issues

1. Use atomic operations to increase and decrease counter.
2. Style fixes.
Remove unused function declaration.

RonEld commented:
@Patater I believe I have addressed your comments

Patater reviewed (commented):
Code LGTM. Still need another comment to setup and teardown implementers to let them know that teardown may be called multiple times, or change the call to only happen when reference_count is `0` (instead of `<= 0`).

RonEld commented:
teardown should not be called multtiple times, it was a check as <= 0 as a corner case.
See https://github.com/ARMmbed/mbed-os/pull/7099#discussion_r192667729

Commit:
Change the terminate limit check

Check for counter to be `< 1` instead of `<= 0` before terminating.

RonEld commented:
@Patater I have changed the counter check before termination

Patater reviewed (approved):
If teardown is called multiple times (sequentially) after the count is 0, terminate will be called multiple times. As discussed, it's OK for terminate to be called multiple times.

RonEld commented:
@Patater Thanks for reviewing.
As discussed, it is OK to call terminate multiple times for the current target using this API.
It might affect other future platforms in need for this function, but this shouldn't block the current PR, as there is only one target using this function at the moment

RonEld commented:
@cmonr @0xc0170 Could you please start CI as this PR was approved?

0xc0170 commented:
> @cmonr @0xc0170 Could you please start CI as this PR was approved?

We will review and run CI (this is expected to be in few days, currently 5.10 PR in the queue)

RonEld commented:
@0xc0170 
>currently 5.10 PR in the queue

To my understanding, this PR is targeting 5.10

simonbutcher reviewed (changes requested):
Minor comments only. My only real concern is your use of the word 'blocker' which I think is misleading.

RonEld commented:
@sbutcher-arm Thank you for your comments. Please refer me to where I use the word "blocker" so i will rephrase.

Commit:
Rename parameter name

Rename `obsolete_ctx` to `unused_ctx` as it is simply unused.

RonEld commented:
@sbutcher-arm i renamed the parameter name to `unused_ctx`

0xc0170 commented:
/morph build

Commit:
Fix build error on IAR

IAR fails to build when a variable is initialized with empty curly braces.
Added `{ { 0 } }` to fix that.

RonEld commented:
build failure on IAR was fixed

NirSonnenschein commented:
/morph build

0xc0170 commented:
Waiting for the last approval from @sbutcher-arm
